WIP - F# compiler guide

cartermp opened the pull request:
This brings the F# compiler technical guide and associated F# compiler service documentation into a new section aimed at giving people a technical dive into the F# compiler and the F# compiler service. The primary audience is those who are interested in contributing to the F# compiler and tools, but it also serves as a technical document for how a modern compiler works.

Also fixes https://github.com/dotnet/docs/issues/10097

cartermp committed:
Initial docs import

cartermp committed:
Add to TOC

cartermp commented:
This is still heavily in progress, as there is much cleanup remaining and multiple sections need updating, especially:

* Notes on performance
* .NET Framework-specific information
* How the .NET Core F# compiler does a few things differently (e.g., assembly resolution, the build)

I can handle most of the performance stuff, but others would need input from @dsyme and @KevinRansom.

@baronfel and @Krzysztof-Cieslak this may also be of interest

cartermp commented:
Also @vzarytovskii and @auduchinok FYI

dsyme commented:
Oh this is great!

I'd always thought this would land in dotnet/fsharp, but am happy to have it on the .NET docs

Krzysztof-Cieslak commented:
Is the goal here to kill docs we host for FCS right now (http://fsharp.github.io/FSharp.Compiler.Service/)?

If so
1. Make sure all docs present there are ported
2. They include *good* Japanese translation ( rather than automatic translations that are used for .NET docs)

dsyme commented:
> Is the goal here to kill docs we host for FCS right now (http://fsharp.github.io/FSharp.Compiler.Service/)?

Good question 

Right now I think it's bringing over https://fsharp.github.io/2015/09/29/fsharp-compiler-guide.html

dsyme commented:
Ah no it's also FCS I see

dsyme commented:
1. The compiler technical guide - yes.  Though it's so technical why not dotnet/fsharp?  With a good landing page on dotnet/docs for the F# compiler?

2. I do feel a bit quesy about taking the FCS docs out of a repository where they are checked and processed against the software.  They are showing how to use an API.  These sort of docs are impossible to maintain in a separate repo and without string checking - I wouldn't even want to try to do that - and this API will certainly evolve further.  

   I guess I don't think it will stand the test of time to have them in dotnet docs.  

3. Agreed we shouldn't lose the multi-language docs for FCS if the community have contributed them?

Krzysztof-Cieslak commented:
I'd imagine compiler content hosted both on https://fsharp.github.io/ and http://fsharp.github.io/FSharp.Compiler.Service/ should get unified in someplace given that there's a plan for having single repo/releasing FCS out of dotnet/fsharp

But I'm not sure if .NET docs are good place for those docs - mostly due to lack of static checking of snippets, more difficult contribution/slower release cycle, and potentially loss of the multi-language docs.

I think we should host those docs as part dotnet/fsharp

dsyme commented:
> I think we should host those docs as part dotnet/fsharp

Yup updated my comment to say the same thing.  Though a landing page for "The F# Compiler" of "F# Tooling" on dotnet/docs would be great (and could link to Ionide, Fable, WebSharper too)

cartermp commented:
A primary goal here is SEO - docs.ms has superior SEO and if we want more eyes on guides for how this stuff works, this is a good way forward. I don't think it means any other docs need to go away. This can be purely additive.

The docs are translated and there is a way to port over a Japanese version that's been hand-translated. It's not a problem.

This PR brings over the FCS docs that are referred to in the compiler guide. Those sections could simply be removed if you like. I'm also unsure why checking of snippets is being considered here. There are no snippets being checked in the docs that I've updated here.

Ultimately these don't need to be here, but there are some downsides to _not_ being a part of a major documentation system to consider.

Krzysztof-Cieslak commented:
I don't think fairly specialized content (compiler guide/FCS guide/etc) needs to be prioritized for SEO - developers that will use it will already know that they should look for it inappropriate place anyway, especially if we'd put it in dotnet/fsharp and link in readme.

So I don't believe it should be our priority here. And we discuss snippets in context of my original question `Is the goal here to kill docs we host for FCS right now` - if that's the plan, typechecking  of snippets is really important discussion point (more important than SEO, in my view)

dsyme commented:
I think the goal is good compiler/tooling content for F# on dotnet/docs?

> I'm also unsure why checking of snippets is being considered here. There are no snippets being checked in the docs that I've updated here.

Ah right, I see, it's not the full literate FCS docs, just a couple of articles.

@Krzysztof-Cieslak That answers part of your question - it's not all of FCS technical docs, so the goal is not to kill those docs

dsyme commented:
Just to say I'd be pretty excited about 

1. having `F# Tooling` content on dotnet/docs

2. having a `docs` directory in `dotnet/fsharp` (hence https://dotnet.github.io/fsharp) that offered good maintainable checked literate content in the standard fashion, plus FCS generated docs there, both auto-published from master

dsyme commented:
> A primary goal here is SEO..

I think there's a related factor of "bringing credibility to participating in F# tooling".  

The F# presence on docs.microsoft.com is really good and plenty of F# traffic heads there.  I think it would be wonderful if people just naturally saw "F# Tooling" and started to learn how to contribute, with links to the major tooling efforts

cartermp commented:
So I think there's no harm at all in not including FCS docs here. If those make the most sense to just be hosted via the FCS site, I've no issue with that. But the sections of the compiler guide that reference those docs need to either link out (trivial), be removed if we want this to be solely about the F# compiler, or adjusted to state what the FCS docs state but purely from the perspective of the compiler. I initially chose to include them here to preserve most of the content, but I've got no issues changing that.

Another purpose here - which I want to include in an index of some sort - is to further emphasize that F# is OSS and you can meaningfully contribute to it. Right now the docs are more "this is a product" and that gives off a bit of a message that's more about consumption rather than consumption + contribution.

Ultimately if this whole thing makes the most sense to host on GitHub I'll change it to do that. I just think that having more of a presence of the OSS nature of F# in these docs is a benefit, and there are some technical aspects of hosting docs here that are wonderful.

dsyme commented:
> Another purpose here - which I want to include in an index of some sort - is to further emphasize that F# is OSS and you can meaningfully contribute to it. 
> 
> I just think that having more of a presence of the OSS nature of F# in these docs is a benefit, and there are some technical aspects of hosting docs here that are wonderful.

I love these goals

Youssef1313 commented on docs/fsharp/tutorials/fsharp-compiler/fsharp-compiler-guide.md:
```diff
@@ -0,0 +1,620 @@
+---
+title: F# compiler guide
+description: Learn about the F# compiler, its various phases, and terminology in the F# compiler codebase.
+ms.date: 8/6/2020
```
```suggestion
ms.date: 08/06/2020
```

Youssef1313 commented on docs/fsharp/tutorials/fsharp-compiler/fsharp-compiler-guide.md:
```diff
@@ -0,0 +1,620 @@
+ms.date: 8/6/2020
+---
+
+# F# Compiler Guide
```
```suggestion
# F# compiler guide
```

Youssef1313 commented on docs/fsharp/tutorials/fsharp-compiler/fsharp-compiler-guide.md:
```diff
@@ -0,0 +1,620 @@
+
+# F# Compiler Guide
+
+This guide discusses the F# Compiler source code and implementation from a technical point of view. The source code for the F# compiler and tools can be found at [`dotnet/fsharp` on GitHub.](https://github.com/dotnet/fsharp).
```
```suggestion
This guide discusses the F# compiler source code and implementation from a technical point of view. The source code for the F# compiler and tools can be found at [`dotnet/fsharp` on GitHub](https://github.com/dotnet/fsharp).
```

Youssef1313 commented on docs/fsharp/tutorials/fsharp-compiler/fsharp-compiler-guide.md:
```diff
@@ -0,0 +1,620 @@
+The F# compiler repositories are used to produce a range of different artifacts.  For the purposes of this
+guide, the important ones are:
+
+### Key Data Formats and Representations 
```
```suggestion
### Key data formats and representations
```

Youssef1313 commented on docs/fsharp/tutorials/fsharp-compiler/fsharp-compiler-guide.md:
```diff
@@ -0,0 +1,620 @@
+
+* _Input source files_  Read as Unicode text, or binary for referenced assemblies.
+
+* _Input command line arguments_  See [CompileOptions.fs](https://github.com/dotnet/fsharp/blob/master/src/fsharp/CompileOptions.fs) for the full  code implementing the arguments table.  Command line arguments are also accepted by the F# Compiler Service API in project specifications, and as optional input to F# Interactive.
```
```suggestion
* _Input command-line arguments_  See [CompileOptions.fs](https://github.com/dotnet/fsharp/blob/master/src/fsharp/CompileOptions.fs) for the full code implementing the arguments table. Command-line arguments are also accepted by the F# Compiler Service API in project specifications, and as optional input to F# Interactive.
```

Youssef1313 commented on docs/fsharp/tutorials/fsharp-compiler/fsharp-compiler-guide.md:
```diff
@@ -0,0 +1,620 @@
+
+* _Tokens_, see [pars.fsy](https://github.com/dotnet/fsharp/blob/master/src/fsharp/pars.fsy), [lex.fsl](https://github.com/dotnet/fsharp/blob/master/src/fsharp/lex.fsl), [lexhelp.fs](https://github.com/dotnet/fsharp/blob/master/src/fsharp/lexhelp.fs) and related files.
+
+* _Abstract Syntax Tree (AST)_, see [SyntaxTree.fs](https://github.com/dotnet/fsharp/blob/master/src/fsharp/SyntaxTree.fs), the untyped syntax tree resulting from parsing
```
```suggestion
* _Abstract Syntax Tree (AST)_, see [SyntaxTree.fs](https://github.com/dotnet/fsharp/blob/master/src/fsharp/SyntaxTree.fs), the untyped syntax tree resulting from parsing.
```

Youssef1313 commented on docs/fsharp/tutorials/fsharp-compiler/fsharp-compiler-guide.md:
```diff
@@ -0,0 +1,620 @@
+
+* _Type checking context/state_, see for example [`TcState` in CompileOps.fs](https://github.com/dotnet/fsharp/blob/master/src/fsharp/CompileOps.fsi) and its constituent parts, particularly `TcEnv` in [TypeChecker.fs](https://github.com/dotnet/fsharp/blob/master/src/fsharp/TypeChecker.fsi) and `NameResolutionEnv` in [NameResolution.fs](https://github.com/dotnet/fsharp/blob/master/src/fsharp/NameR ...
+
+* _Abstract IL_, the output of code generation, then used for binary generation, and the input format when reading .NET assemblies, see [`ILModuleDef` in il.fs](https://github.com/dotnet/fsharp/blob/master/src/absil/il.fsi)
```
```suggestion
* _Abstract IL_, the output of code generation, then used for binary generation, and the input format when reading .NET assemblies, see [`ILModuleDef` in il.fs](https://github.com/dotnet/fsharp/blob/master/src/absil/il.fsi).
```

Youssef1313 commented on docs/fsharp/tutorials/fsharp-compiler/fsharp-compiler-guide.md:
```diff
@@ -0,0 +1,620 @@
+
+* _The .NET Binary format_ (with added "pickled" F# Metadata resource), the final output of fsc.exe, see the ECMA 335 specification and the [ilread.fs](https://github.com/dotnet/fsharp/blob/master/src/absil/ilread.fs) and [ilwrite.fs](https://github.com/dotnet/fsharp/blob/master/src/absil/ilwrite.fs) binary reader/generator implementations. The added F# metadata is stored in a binary resource, se ...
+
+* _The incrementally emitted .NET reflection assembly,_ the incremental output of fsi.exe. See [ilreflect.fs](https://github.com/dotnet/fsharp/blob/master/src/absil/ilreflect.fs)
```
```suggestion
* _The incrementally emitted .NET reflection assembly,_ the incremental output of fsi.exe. See [ilreflect.fs](https://github.com/dotnet/fsharp/blob/master/src/absil/ilreflect.fs).
```

Youssef1313 commented on docs/fsharp/tutorials/fsharp-compiler/fsharp-compiler-guide.md:
```diff
@@ -0,0 +1,620 @@
+
+END TODO
+
+### Key Compiler Phases
```
```suggestion
### Key compiler phases
```

Youssef1313 commented on docs/fsharp/tutorials/fsharp-compiler/fsharp-compiler-guide.md:
```diff
@@ -0,0 +1,620 @@
+
+### Key Compiler Phases
+
+Below, you can see a diagram of how different phases of F# compiler work:
```
```suggestion
The following is a diagram of how different phases of F# compiler work:
```

Youssef1313 commented on docs/fsharp/tutorials/fsharp-compiler/fsharp-compiler-guide.md:
```diff
@@ -0,0 +1,620 @@
+* _Abstract IL code rewriting_, see [EraseClosures.fs](https://github.com/dotnet/fsharp/blob/master/src/ilx/EraseClosures.fs) and
+  [EraseUnions.fs](https://github.com/dotnet/fsharp/blob/master/src/ilx/EraseUnions.fs). Eliminates some constructs by rewriting Abstract IL nodes.
+  
+* _Binary emit_, see [ilwrite.fsi](https://github.com/dotnet/fsharp/blob/master/src/absil/ilwrite.fsi)/[ilwrite.fs](https://github.com/dotnet/fsharp/blob/master/src/absil/ilwrite.fs). 
```
```suggestion
* _Binary emit_, see [ilwrite.fsi](https://github.com/dotnet/fsharp/blob/master/src/absil/ilwrite.fsi)/[ilwrite.fs](https://github.com/dotnet/fsharp/blob/master/src/absil/ilwrite.fs).
```

Youssef1313 commented on docs/fsharp/tutorials/fsharp-compiler/fsharp-compiler-guide.md:
```diff
@@ -0,0 +1,620 @@
+  
+* _Binary emit_, see [ilwrite.fsi](https://github.com/dotnet/fsharp/blob/master/src/absil/ilwrite.fsi)/[ilwrite.fs](https://github.com/dotnet/fsharp/blob/master/src/absil/ilwrite.fs). 
+
+* _Reflection-Emit_, see [ilreflect.fs](https://github.com/dotnet/fsharp/blob/master/src/absil/ilreflect.fs). 
```
```suggestion
* _Reflection-Emit_, see [ilreflect.fs](https://github.com/dotnet/fsharp/blob/master/src/absil/ilreflect.fs).
```

Youssef1313 commented on docs/fsharp/tutorials/fsharp-compiler/fsharp-compiler-guide.md:
```diff
@@ -0,0 +1,620 @@
+
+* _The F# Compiler Shell_, see [fsc.fs](https://github.com/dotnet/fsharp/blob/master/src/fsharp/fsc.fs) and [fscmain.fs](https://github.com/dotnet/fsharp/blob/master/src/fsharp/fscmain.fs).
+
+## Coding Standards and Idioms
```
```suggestion
## Coding standards and idioms
```

Youssef1313 commented on docs/fsharp/tutorials/fsharp-compiler/fsharp-compiler-guide.md:
```diff
@@ -0,0 +1,620 @@
+
+When formatting "info" objects, see the functions in the ``NicePrint`` module.
+
+
```
```suggestion
```

Youssef1313 commented on docs/fsharp/tutorials/fsharp-compiler/fsharp-compiler-guide.md:
```diff
@@ -0,0 +1,620 @@
+
+### Notes on displaying types
+
+When displaying a type, you will normally want to "prettify" the type first (i.e. make the type "pretty"). 
```
```suggestion
When displaying a type, you will normally want to "prettify" the type first (make the type "pretty").
```

Youssef1313 commented on docs/fsharp/tutorials/fsharp-compiler/fsharp-compiler-service-caches.md:
```diff
@@ -0,0 +1,42 @@
+
+* `scriptClosureCache` - an MRU cache of default size `projectCacheSize` that caches the computation of GetProjectOptionsFromScript. This computation can be lengthy as it can involve processing the transitive closure of all `#load` directives, which in turn can mean parsing an unbounded number of script files.
+
+* `incrementalBuildersCache` - an MRU cache of projects where a handle is being kept to their incremental checking state, of default size `projectCacheSize` (200 by default, unless explicitly set as a parameter). The "current background project" (see the [FSharpChecker operations queue](queue.html)) will be one of these projects. When analyzing large collections of projects, this cache usually oc ...
```
queue.html is invalid link.

Youssef1313 commented on docs/fsharp/tutorials/fsharp-compiler/fsharp-compiler-service-queue.md:
```diff
@@ -0,0 +1,29 @@
+ms.date: 8/6/2020
+---
+
+## F# Compiler Service Queue
```
```suggestion
# F# Compiler service queue
```

cartermp committed:
Apply suggestions from code review

Co-authored-by: Youssef Victor <31348972+Youssef1313@users.noreply.github.com>

cartermp committed:
More cleanup, cover optimizations more thoroughly

cartermp commented:
I'm going to keep the two FCS documents in here while I flesh out the documentation, but I plan on removing them and linking to FCS docs proper (with a contribution to FCS docs for these articles). It's handy to have them present right now since I'll likely be making some changes as I go through the tooling-specific things, especially as it pertains to performance.

cartermp committed:
markdownlint

cartermp committed:
markdownlint

cartermp committed:
clarify more optimizations

cartermp committed:
Start to rewrite perf section

cartermp committed:
perf section revamp

cartermp committed:
more updates

cartermp commented:
@dsyme I left some TODOs in the text here specifically for you. Would you mind going through them?

Separately, I:

* Rewrote the perf section, focusing more on general concepts and techniques and removed the out of date data
* Revamped and reworded some various things
* Updated all links and names of concepts
* Removed lots of outdated info
* Added more examples and descriptions of optimization passes

Each can be improved but overall the document is in much better shape now.

The big one left is the discussion about resource files. I don't know how to handle that one, aside from removing some mentions of differences with C# that are no longer the case.

dsyme commented:
> @dsyme I left some TODOs in the text here specifically for you. Would you mind going through them?

It's on my todo list, hopefully tomorrow

dsyme commented on docs/fsharp/tutorials/fsharp-compiler/fsharp-compiler-guide.md:
```diff
@@ -0,0 +1,670 @@
+
+* _The F# Compiler Shell_, see [fsc.fs](https://github.com/dotnet/fsharp/blob/master/src/fsharp/fsc.fs) and [fscmain.fs](https://github.com/dotnet/fsharp/blob/master/src/fsharp/fscmain.fs).
+
+## Coding standards and idioms
```
I don't think this table should be in these docs, which should be a taster and contain nothing that may rot over time.

Instead we should start a `docs` directory in dotnet/fsharp and put this table there (plus a copy of most of the above information)

dsyme commented on docs/fsharp/tutorials/fsharp-compiler/fsharp-compiler-guide.md:
```diff
@@ -0,0 +1,670 @@
+| `IL`                 | Abstract  IL = F# representation of .NET IL |
+| `Ilx`                 | Extended Abstract IL = .NET IL plus a coulpe of contructs that get erased |
+
+## Adding Error Messages
```
Likewise these belong in DEVNOTES.md or docs directory of dotnet/fsharp

dsyme commented on docs/fsharp/tutorials/fsharp-compiler/fsharp-compiler-guide.md:
```diff
@@ -0,0 +1,670 @@
+* _The F# Compiler Shell_, see [fsc.fs](https://github.com/dotnet/fsharp/blob/master/src/fsharp/fsc.fs) and [fscmain.fs](https://github.com/dotnet/fsharp/blob/master/src/fsharp/fscmain.fs).
+
+## Coding standards and idioms
+
```
In fact I've been through all the information below and to be honest I think all of the rest should go in dotnet/fsharp.  docs/dotnet is not thr right place to maintain such highly technical information 

What's in the first section is ok.  But it should only be the slice we need to give a taster and which will not change at all.  Otherwise we are creating a doc set that will not stand the test of time.

dsyme commented on docs/fsharp/tutorials/fsharp-compiler/fsharp-compiler-guide.md:
```diff
@@ -0,0 +1,670 @@
+
+More generally, any time a first-order function is passed as an argument to a second-order function, the first-order function is not inlined even if everything is marked as `inline`. This results in a performance penalty.
+
+## Compiler Startup Performance
```
This section could perhaps stay (the topic is good), though I'd recommend a very short section on dotnet/docs pointing to a longer living document in dotnet/fsharp docs

dsyme commented on docs/fsharp/tutorials/fsharp-compiler/fsharp-compiler-guide.md:
```diff
@@ -0,0 +1,670 @@
+
+`FSharp.Build.dll` and `Microsoft.FSharp.targets` give MSBuild support for F# projects (`.fsproj`). Although not strictly part of the F# compiler, these components are always found in F# tool distributions for Mono and Windows.
+
+### How F# handles EmbeddedResource items differently to C# (details)
```
This should be removed completely

dsyme commented on docs/fsharp/tutorials/fsharp-compiler/fsharp-compiler-service-queue.md:
```diff
@@ -0,0 +1,28 @@
+---
```
Move to `docs` directory in dotnet/fsharp

dsyme reviewed (changes requested):
Hi @cartermp , I've left some comments - mostly saying that the majority (but not all) should move to a docs directory on dotnet/fsharp.  I'm very motivated to help make such a directory merging the FCS docs and all other information sources we have.

baronfel commented:
For what it's worth, the FCS docs are almost buildable using FSharp.Formatting: https://github.com/baronfel/fsharp/tree/fcs-docs.

The major missing feature in FSharp.Formatting is multi-language support for the generated pages, which will be a blocker to completing the work.

dsyme commented:
> The major missing feature in FSharp.Formatting is multi-language support for the generated pages, which will be a blocker to completing the work.

Could you add an FSHarp.Formatting issue describing what we think we need? https://github.com/fsprojects/FSharp.Formatting/

Presumably https://github.com/baronfel/fsharp/tree/fcs-docs/docs/fcs/docs/ja all get processed already?  What else is needed?

dsyme commented:
In https://github.com/baronfel/fsharp/tree/fcs-docs/docs I think just move all of those docs into `docs\service`?

baronfel commented:
> Could you add an FSHarp.Formatting issue describing what we think we need?

yep! I was collecting screenshots and descriptions of what I think the gaps are and what other tools do here.
